File: Tetris.py
import os, time, random, readchar, copy, sys, queue, threading
from datetime import datetime

#program requires pip and readchar library from pip
#game was written without using OOP method
#using 'class' as public method (or be called 'struct' in different language)
#the game are created for Ubuntu OS only

#init board's size
colSize = 10
rowSize = 20

#queue to manage threads
inputQueue = queue.Queue()

#init global array as square matrix, we will cut unneccesary parts of matrix (i.e row or column with fully '0') later
shapeO = [
    [1,1],
    [1,1]
]
shapeL = [
    [0,0,1],
    [1,1,1],
    [0,0,0]
]
shapeJ = [
    [1,0,0],
    [1,1,1],
    [0,0,0]
]
shapeS = [
    [0,1,1],
    [1,1,0],
    [0,0,0]
]
shapeZ = [
    [1,1,0],
    [0,1,1],
    [0,0,0]
]
shapeI = [
    [0,0,0,0],
    [1,1,1,1],
    [0,0,0,0],
    [0,0,0,0]
]
shapeT = [
    [0,1,0],
    [1,1,1],
    [0,0,0]
]

#initialize 3-Dimensional array contain various type of block's shape
blockShape = [shapeO,shapeL,shapeJ,shapeS,shapeZ,shapeI,shapeT]

#initialize color code array for display colorful block
#we don't use black color due to black theme Terminal are
colorCode = [
	"\033[0m  \033[m",          #blank
	"\033[0;31;41m[]\033[m",    #Text: Red, Background: Red
	"\033[0;32;42m[]\033[m",    #Text: Green, Background: Green
	"\033[0;33;43m[]\033[m",    #Text: Yellow, Background: Yellow
	"\033[0;34;44m[]\033[m",    #Text: Blue, Background: Blue
	"\033[0;35;45m[]\033[m",    #Text: Purple, Background: Purple
	"\033[0;36;46m[]\033[m",    #Text: Cyan, Background: Cyan
	"\033[0;37;47m[]\033[m"     #Text: White, Background: White
]

#SRS rotation
#Wall Kick case
#apply for J,L,Z,S,T Tetromino's case following clockwise direction
#test1 will be all {0,0} so we will skip test1
case1 = [[-1, 0], [-1, 1], [0, -2], [-1, -2]] #0->R     #R means Right
case2 = [[1, 0], [1, -1], [0, 2], [1, 2]]     #R->2     #2 means succeed 2nd rotation
case3 = [[1, 0], [1, 1], [0, -2], [1, -2]]    #2->L     #L means Left
case4 = [[-1, 0], [-1, -1], [0, 2], [-1, 2]]  #L->0
#for I Tetromino's case (clockwise direction)
case1_i = [[-2, 0], [1, 0], [-2, -1], [1, 2]] #0->R
case2_i = [[-1, 0], [2, 0], [-1, 2], [2, -1]] #R->2
case3_i = [[2, 0], [-1, 0], [2, 1], [-1, -2]] #2->L
case4_i = [[1, 0], [-2, 0], [1, -2], [-2, 1]] #L->0

#initialize block as an Object (including elements: coordinate 'x':int, coordinate 'y':int, shape of block:[][]int)
#all element of object 'Block' are public, which means you can call it everytime you want
#it will be easier to work when using with coordinates
class Block(object):
    def __init__(self,x,y,shape,shapeType,rotateType):
        self.x = x              #storing horizontal coordinate x
        self.y = y              #storing vertical coordinate y
        self.shape = shape[:]   #shape[][]:int - storing shape of block
        self.shapeType = shapeType  #store following index of blockShape (range [0,6] means shapeO->shapeT above). For rotation supporting
        self.rotateType = rotateType    #store 1 number in range [0,3]. For rotation supporting
    #change all value 1 to another value in range [1,7] in order to get various colorful block 
    def reInit(self):
        num = random.randint(1,7)
        for i in range(len(self.shape)):
            for j in range(len(self.shape[0])):
                if self.shape[i][j] != 0:
                    self.shape[i][j] = num
    #transpose 2-Dimensional (square) matrix
    def transpose(self):
        for i in range(len(self.shape)): 
            for j in range(i+1,len(self.shape[0])):
                self.shape[i][j], self.shape[j][i] = self.shape[j][i], self.shape[i][j]
    #reverse (swap) each layer column of 2D matrix
    #input: arr[][]:int
    #output: arr[][]:int
    #reference: https://stackoverflow.com/questions/42519/how-do-you-rotate-a-two-dimensional-array/8664879#8664879?newreg=e1f2cab8974b45dcbe5ac61e8d1be2c9
    def reverseColumn(self):
        for i in range(len(self.shape)):
            for j in range(int(len(self.shape[0])/2)):
                self.shape[i][j], self.shape[i][len(self.shape[0])-1-j] = self.shape[i][len(self.shape[0])-1-j], self.shape[i][j]
    #rotate block
    def rotate(self,land):
        previousShape = copy.deepcopy(self.shape)
        self.transpose()
        self.reverseColumn()
        tempX = self.x
        tempY = self.y
        flag = False
        #if collision happen after the rotation
        #do all possible test in caseX array
        #X is number value in range [0,3] depends on rotateType, which shows transformation of block e.g 0->R, R->2)
        if checkCollision(land,self):
            #if block is not I Tetromino
            if self.shapeType != 5:
                if self.rotateType == 0:    #0->R
                    for i in range(len(case1)):
                        self.x += case1[i][0]
                        self.y += case1[i][0]
                        if not checkCollision(land,self):
                            flag = True
                            break
                        else:
                            self.x = tempX
                            self.y = tempY
                elif self.rotateType == 1:  #R->2
                    for i in range(len(case2)):
                        self.x += case2[i][0]
                        self.y += case2[i][0]
                        if not checkCollision(land,self):
                            flag = True
                            break
                        else:
                            self.x = tempX
                            self.y = tempY
                elif self.rotateType == 2:  #2->L
                    for i in range(len(case3)):
                        self.x += case3[i][0]
                        self.y += case3[i][0]
                        if not checkCollision(land,self):
                            flag = True
                            break
                        else:
                            self.x = tempX
                            self.y = tempY
                else:   #== 3   #L->0
                    for i in range(len(case4)):
                        self.x += case4[i][0]
                        self.y += case4[i][0]
                        if not checkCollision(land,self):
                            flag = True
                            break
                        else:
                            self.x = tempX
                            self.y = tempY
            #if block is I Tetromino
            else:   #case shape I
                if self.rotateType == 0:    #0->R
                    for i in range(len(case1_i)):
                        self.x += case1_i[i][0]
                        self.y += case1_i[i][0]
                        if not checkCollision(land,self):
                            flag = True
                            break
                        else:
                            self.x = tempX
                            self.y = tempY
                elif self.rotateType == 1:  #R->2
                    for i in range(len(case2_i)):
                        self.x += case2_i[i][0]
                        self.y += case2_i[i][0]
                        if not checkCollision(land,self):
                            flag = True
                            break
                        else:
                            self.x = tempX
                            self.y = tempY
                elif self.rotateType == 2:  #2->L
                    for i in range(len(case3_i)):
                        self.x += case3_i[i][0]
                        self.y += case3_i[i][0]
                        if not checkCollision(land,self):
                            flag = True
                            break
                        else:
                            self.x = tempX
                            self.y = tempY
                else:   #== 3   #L->0
                    for i in range(len(case4_i)):
                        self.x += case4_i[i][0]
                        self.y += case4_i[i][0]
                        if not checkCollision(land,self):
                            flag = True
                            break
                        else:
                            self.x = tempX
                            self.y = tempY
        else:
            flag = True
        #check if rotate was succeed or not
        if flag == False:   #failed
            self.shape = previousShape
        else:               #succeed
            if self.rotateType == 3:
                self.rotateType = 0
            else:
                self.rotateType += 1
        
    def moveDown(self,land):
        self.y += 1
        if checkCollision(land,self):
            self.y -= 1

    def moveLeft(self,land):
        self.x -= 1
        if checkCollision(land,self):
            self.x += 1

    def moveRight(self,land):
        self.x += 1
        if checkCollision(land,self):
            self.x -= 1

    def hardDrop(self,land):
        while not checkCollision(land,self):
            self.y += 1
        #we make it into not checkCollision after force to drop it until collision happen
        self.y -= 1

#be careful with initializing 2-Dimensional array
#initialize board (new board)
#input:
#output: board[][]:int
def initBoard():
    board = [[0 for j in range(colSize)] for i in range(rowSize)]
    return board

#get random block
#input: shape[][][]:int (storing all shape of block), board[][]:int
#output: block:Block(object)
def randomBlock(shape,land):
    randomNumber = random.randint(0,6)
    randomShape = copy.deepcopy(shape[randomNumber])
    #block will display outside of land (more detail: block location will be higher than roof of land due to "-2" coordinate 'y')
    block = Block(int(len(land[0])/2) - len(randomShape[0]) + 1, -2, randomShape, randomNumber, 0)      #block will appear at the middle of the top of the board
    block.reInit()
    return block

#merge block with board (temporary board)
#input: board[][]:int, block:Block(object)
#output: board[][]:int
def mergeBlock(land,block):
    #create temporary variable to avoid change of array's elements' values
    temp = copy.deepcopy(land)
    for i in range(len(block.shape)):
        for j in range(len(block.shape[0])):
            if block.shape[i][j] != 0:
                #additional condition for display block even block is outside (above, higher from roof of land)
                if block.y + i <= rowSize - 1 and block.y + i >= 0 and block.x + j <= colSize - 1 and block.x + j >= 0:
                    temp[block.y+i][block.x+j] = block.shape[i][j]
    return temp

#collision between block and land (main board), between block and pieces of other previous block
#also the most important function
#input: land[][]:int, block:Block(object)
#output: True or False
def checkCollision(land,block):
    for i in range(len(block.shape)):
        for j in range(len(block.shape[0])):
            if block.shape[i][j] != 0:
                if block.x+j < 0 or block.x+j > colSize-1 or block.y+i > rowSize-1:
                    return True
                elif block.x+j >= 0 and block.x+j <= colSize - 1 and block.y+i >= 0 and block.y+i <= rowSize-1:
                    if land[block.y+i][block.x+j] != 0:
                        return True
    return False

#reach top??? we just need to check the first line at the top of the land
#if it has any piece, return True to alarm that the game will be over at the next move 
#input: land[][]:int
#output: True or False
def checkReachTop(land):
    for j in range(len(land[0])):
        if land[0][j] != 0:
            return True         #the game will be over
    return False                #the game is continuing
        
#check every single line of land array (main board)
#we will delete lines which is full by make it clear with 0 and swap very couple lines from i to 0 as (line[i] = line[i-1]), (line[i-1] = line[i-2]), ...
#swap from under line to top line 
#input: land[][]:int, score:int
#output: land[][]:int (after checking and re-merging), score:int 
def handleFullLine(land,score):
    flag = True
    line = []
    #collect all line was full
    for i in range(len(land)):
        flag = True
        for j in range(len(land[0])):
            if land[i][j] == 0:
                flag = False
                break
        if flag == True:
            line.append(i)
    
    for i in line:
        score += 100    #get score
        for j in range(len(land[0])):
            land[i][j] = 0
        #copy land's row which is clear with value 0
        temp = copy.deepcopy(land[i])
        #take that row out
        land.pop(i)
        #and insert it to the first position of array 'land'
        land.insert(0,temp)

    return land, score

#delay program for t second
#input: t:int
def sleep(t):
    time.sleep(t)

#clear console
def clearScreen():
    os.system("clear")

#in: arr:[][]int, nextblock:Block(object), score:int, color:[]string
#out: []string
def outputConvert(land,block,nextblock,score,color):
    arr = mergeBlock(land,block)
    #temporary variables for printing out next Block
    tempIndexRow = 0
    tempIndexCol = 0
    #output array type string
    strout = []
    tempstr = ""
    for i in range(len(arr)):
        #refresh
        tempIndexCol = 0
        tempstr = ""
        #
        if i == 0:
            strout.append("  " + "_ "*colSize)
        for j in range(len(arr[0])):
            if j == 0:
                tempstr += "||"
            if i == len(arr)-1 and arr[i][j] == 0:
                tempstr += "_ "
            else:
                tempstr += color[arr[i][j]]
            if j == len(arr[0])-1:
                tempstr += "||"
        if i == 1:
            tempstr += "   Score: " + str(score)
        elif i == 3:
            tempstr += "   Next"
        elif i > 4 and tempIndexRow < len(nextblock.shape):
            tempstr += "   "
            while tempIndexCol < len(nextblock.shape[0]):
                tempstr += color[nextblock.shape[tempIndexRow][tempIndexCol]]
                tempIndexCol += 1
            tempIndexRow += 1
         #   
        strout.append(tempstr)

    return strout 

#print out 2D array and score
#input: arr[][]:int, nextBlock:Block(object), score:int, color:[]string
def display(strout):
    clearScreen()
    for i in range(len(strout)):
        print("\r" + strout[i])
    print("\rPress: W - Rotate, S - Fall down faster, A - Move left, D - Move Right")
    print("\rPress: Spacebar - Hard Drop")
    print("\rPress: X to exit the program ")

#thread get input from keyboard
def getKey():
    while True:
        key = readchar.readkey()
        inputQueue.put(key)
        if key == "x" or key == "X":
            sys.exit()

#do movement according to input
#input: board, land:[][]int; block:Block(object); inputQueue:Queue; score:int; levelUp:int; color:[]string
#output: board, land:[][]int; block:Block(object); inputQueue:Queue; key:string
def doMovement(land,block,nextBlock,intputQueue,score,levelUp,color):
    key = ""
    preSec = int(round(time.time()*1000))
    while not inputQueue.empty():
        key = inputQueue.get()
        if key == "w" or key == "W":
            block.rotate(land)
        elif key == "a" or key == "A":
            block.moveLeft(land)
        elif key == "d" or key == "D":
            block.moveRight(land)
        elif key == "s" or key == "S":
            block.moveDown(land)
        elif key == " ":
            block.hardDrop(land)
        elif key == "x" or key == "X":
            break
        else:
            continue
        performArray = outputConvert(land,block,nextBlock,score,color)
        display(performArray)
        sleep(0.01)
        #do movement in only 1 second (can be decreased due to increasing of scores)
        if int(round(time.time()*1000)) - preSec >= 1000 - levelUp or key == " ":
            break
    return block, inputQueue, key

#input: color:[]string, shape:[][][]int, inputQueue:Queue(object)
def Tetris(color,shape,inputQueue):
    #Main program begin here:
    #Initialization
    land = initBoard()
    score = 0
    levelUp = 0
    preSec = 0
    key = ""
    #
    block = randomBlock(shape,land)
    nextBlock = randomBlock(shape,land)
    #Begin the game, print out the first scene
    performArray = outputConvert(land,block,nextBlock,score,color)
    display(performArray)
    while True:
        levelUp = int(160 * int(score/500))
        if score >= 3000:
            levelUp = 900
        block, inputQueue, key = doMovement(land,block,nextBlock,inputQueue,score,levelUp,color)
        if int(round(time.time()*1000)) - preSec >= 1000-levelUp or key == " ":
            block.y += 1
            if checkCollision(land,block):
                block.y -= 1
                inputQueue.queue.clear()
                land = mergeBlock(land,block)
                land, score = handleFullLine(land,score)
                block = nextBlock
                nextBlock = randomBlock(shape,land)
            preSec = int(round(time.time()*1000))
            performArray = outputConvert(land,block,nextBlock,score,color)
            display(performArray)
        if checkReachTop(land) or key == "x" or key == "X":
            print("\r\033[1;31mGAME OVER\033[m")
            sys.exit()
        else:
            sleep(0.01)

#initialize threads
drive1 = threading.Thread(target=Tetris,args=(colorCode,blockShape,inputQueue))
drive2 = threading.Thread(target=getKey)
#start threads
drive1.start()
drive2.start()

File: Tetris.go
package main

import (
	"fmt"
	"math/rand"
	"os"
	"os/exec"
	"time"
)

var (
	colSize int = 10
	rowSize int = 20
	//
	shapeO = [][]int{
		{1, 1},
		{1, 1}}
	shapeL = [][]int{
		{0, 0, 1},
		{1, 1, 1},
		{0, 0, 0}}
	shapeJ = [][]int{
		{1, 0, 0},
		{1, 1, 1},
		{0, 0, 0}}
	shapeS = [][]int{
		{0, 1, 1},
		{1, 1, 0},
		{0, 0, 0}}
	shapeZ = [][]int{
		{1, 1, 0},
		{0, 1, 1},
		{0, 0, 0}}
	shapeI = [][]int{
		{0, 0, 0, 0},
		{1, 1, 1, 1},
		{0, 0, 0, 0},
		{0, 0, 0, 0}}
	shapeT = [][]int{
		{0, 1, 0},
		{1, 1, 1},
		{0, 0, 0}}
	//
	blockShape = [][][]int{shapeO, shapeL, shapeJ, shapeS, shapeZ, shapeI, shapeT}

	//
	colorCode = []string{
		"\033[0m  \033[m",       //blank
		"\033[0;31;41m[]\033[m", //Text: Red, Background: Red
		"\033[0;32;42m[]\033[m", //Text: Green, Background: Green
		"\033[0;33;43m[]\033[m", //Text: Yellow, Background: Yellow
		"\033[0;34;44m[]\033[m", //Text: Blue, Background: Blue
		"\033[0;35;45m[]\033[m", //Text: Purple, Background: Purple
		"\033[0;36;46m[]\033[m", //Text: Cyan, Background: Cyan
		"\033[0;37;47m[]\033[m"} //Text: White, Background: White
	//SRS rotation
	//Wall Kick case
	//apply for J,L,Z,S,T Tetromino's case following clockwise direction
	//test1 will be all {0,0} so we will skip test1
	case1 = [][]int{{-1, 0}, {-1, 1}, {0, -2}, {-1, -2}} //0->R
	case2 = [][]int{{1, 0}, {1, -1}, {0, 2}, {1, 2}}     //R->2
	case3 = [][]int{{1, 0}, {1, 1}, {0, -2}, {1, -2}}    //2->L
	case4 = [][]int{{-1, 0}, {-1, -1}, {0, 2}, {-1, 2}}  //L->0
	//for I Tetromino's case (clockwise direction)
	case1_i = [][]int{{-2, 0}, {1, 0}, {-2, -1}, {1, 2}} //0->R
	case2_i = [][]int{{-1, 0}, {2, 0}, {-1, 2}, {2, -1}} //R->2
	case3_i = [][]int{{2, 0}, {-1, 0}, {2, 1}, {-1, -2}} //2->L
	case4_i = [][]int{{1, 0}, {-2, 0}, {1, -2}, {-2, 1}} //L->0
)

//
type Block struct {
	x          int
	y          int
	shape      [][]int
	shapeType  int //store following index of blockShape (range [0,6] means shapeO->shapeT above). For rotation supporting
	rotateType int //store 1 number in range [0,3]. For rotation supporting
}

//
func (block *Block) reInit() {
	randomNum := rand.Intn(7-1) + 1
	for i := range block.shape {
		for j := range block.shape[0] {
			if block.shape[i][j] != 0 {
				block.shape[i][j] = randomNum
			}
		}
	}
}

//
func (block *Block) transpose() {
	for i := range block.shape {
		for j := 0 + i; j < len(block.shape[0]); j++ {
			//swap
			block.shape[j][i], block.shape[i][j] = block.shape[i][j], block.shape[j][i]
		}
	}
}

//
func (block *Block) reverseColumn() {
	temp := 0
	pivot := int(len(block.shape[0]) / 2)
	//swap column i with column len(block.shape)-1-i
	for i := range block.shape {
		for j := 0; j < pivot; j++ {
			temp = block.shape[i][j]
			block.shape[i][j] = block.shape[i][len(block.shape[0])-1-j]
			block.shape[i][len(block.shape[0])-1-j] = temp
		}
	}
}

//
func (block *Block) rotate(land [][]int) {
	previousShape := makeCopy(block.shape) //store previouse shape after rotated shape
	block.transpose()
	block.reverseColumn()
	tempX := block.x
	tempY := block.y
	flag := false
	//
	if checkCollision(land, *block) {
		//following my code, index 5 is I Tetromino, which has its own wall kick case
		if block.shapeType != 5 {
			if block.rotateType == 0 { //0->R
				for i := range case1 {
					block.x += case1[i][0]
					block.y += case1[i][1]
					if !checkCollision(land, *block) {
						flag = true
						break
					} else {
						block.x = tempX
						block.y = tempY
					}
				}
			} else if block.rotateType == 1 { //R->2
				for i := range case2 {
					block.x += case2[i][0]
					block.y += case2[i][1]
					if !checkCollision(land, *block) {
						flag = true
						break
					} else {
						block.x = tempX
						block.y = tempY
					}
				}
			} else if block.rotateType == 2 { //2->L
				for i := range case3 {
					block.x += case3[i][0]
					block.y += case3[i][1]
					if !checkCollision(land, *block) {
						flag = true
						break
					} else {
						block.x = tempX
						block.y = tempY
					}
				}
			} else { //case == 3	//L->0
				for i := range case4 {
					block.x += case4[i][0]
					block.y += case4[i][1]
					if !checkCollision(land, *block) {
						flag = true
						break
					} else {
						block.x = tempX
						block.y = tempY
					}
				}
			}
		} else {
			if block.rotateType == 0 { //0->R
				for i := range case1_i {
					block.x += case1_i[i][0]
					block.y += case1_i[i][1]
					if !checkCollision(land, *block) {
						flag = true
						break
					} else {
						block.x = tempX
						block.y = tempY
					}
				}
			} else if block.rotateType == 1 { //R->2
				for i := range case2_i {
					block.x += case2_i[i][0]
					block.y += case2_i[i][1]
					if !checkCollision(land, *block) {
						flag = true
						break
					} else {
						block.x = tempX
						block.y = tempY
					}
				}
			} else if block.rotateType == 2 {
				for i := range case3_i {
					block.x += case3_i[i][0]
					block.y += case3_i[i][1]
					if !checkCollision(land, *block) {
						flag = true
						break
					} else {
						block.x = tempX
						block.y = tempY
					}
				}
			} else { //case == 3	//L->0
				for i := range case4_i {
					block.x += case4_i[i][0]
					block.y += case4_i[i][1]
					if !checkCollision(land, *block) {
						flag = true
						break
					} else {
						block.x = tempX
						block.y = tempY
					}
				}
			}
		}
	} else {
		flag = true
	}
	//if rotation was failed
	if flag == false {
		block.shape = previousShape
		return
	}
	//
	if block.rotateType == 3 {
		block.rotateType = 0
	} else {
		block.rotateType += 1
	}
}

//
func (block *Block) moveDown(land [][]int) {
	block.y++
	if checkCollision(land, *block) {
		block.y--
	}
}

//
func (block *Block) moveLeft(land [][]int) {
	block.x--
	if checkCollision(land, *block) {
		block.x++
	}
}

//
func (block *Block) moveRight(land [][]int) {
	block.x++
	if checkCollision(land, *block) {
		block.x--
	}
}

//
func (block *Block) hardDrop(land [][]int) {
	for !checkCollision(land, *block) {
		block.y++
	}
	block.y--
}

//
func makeCopy(arr [][]int) [][]int {
	newArray := make([][]int, len(arr))
	for i := range arr {
		newArray[i] = make([]int, len(arr[0]))
		for j := range arr[0] {
			newArray[i][j] = arr[i][j]
		}
	}
	return newArray
}

//
func checkCollision(land [][]int, block Block) bool {
	for i := range block.shape {
		for j := range block.shape[0] {
			if block.shape[i][j] != 0 {
				if block.x+j < 0 || block.x+j > colSize-1 || block.y+i > rowSize-1 {
					return true
				} else if block.x+j >= 0 && block.x+j <= colSize-1 && block.y+i <= rowSize-1 && block.y+i >= 0 {
					if land[block.y+i][block.x+j] != 0 {
						return true
					}
				}
			}
		}
	}
	return false
}

//
func randomBlock(board [][]int, Shape [][][]int) Block {
	randomNum := rand.Intn(len(Shape) - 1)
	randomShape := makeCopy(Shape[randomNum])
	coordinateX := int(len(board[0])/2) - len(randomShape[0]) + 1
	coordinateY := -2
	block := Block{coordinateX, coordinateY, randomShape, randomNum, 0}
	block.reInit()
	return block
}

//
func initBoard() [][]int {
	board := make([][]int, rowSize)
	for i := range board {
		board[i] = make([]int, colSize)
	}
	return board
}

//
func mergeBlock(board [][]int, block Block) [][]int {
	for i := range block.shape {
		for j := range block.shape[0] {
			if block.shape[i][j] != 0 {
				if block.y+i >= 0 && block.y+i <= rowSize-1 && block.x+j >= 0 && block.x+j <= colSize-1 {
					board[block.y+i][block.x+j] = block.shape[i][j]
				}
			}
		}
	}
	return board
}

//
func checkReachTop(land [][]int) bool {
	for i := range land[0] {
		if land[0][i] != 0 {
			return true
		}
	}
	return false
}

//
func handleFullLine(land [][]int, score int) ([][]int, int) {
	var line []int
	var flag bool
	for i := range land {
		flag = true
		for j := range land[0] {
			if land[i][j] == 0 {
				flag = false
				break
			}
		}
		if flag == true {
			line = append(line, i)
		}
	}
	//index is line (row), j is column
	for _, index := range line {
		for j := range land[index] {
			land[index][j] = 0
		}
		for j := index; j > 0; j-- {
			//swap
			land[j-1], land[j] = land[j], land[j-1]
		}
		score += 100
	}
	return land, score
}

//
func ClearScreen() {
	cmd := exec.Command("clear")
	cmd.Stdout = os.Stdout
	cmd.Run()
}

//
func Display(arr [][]int, nextBlock Block, score int, Color []string) {
	ClearScreen()
	tempIndexRow := 0
	tempIndexCol := 0
	for i := range arr {
		tempIndexCol = 0
		if i == 0 {
			fmt.Print("\r  ")
			for k := 0; k < len(arr[0]); k++ {
				fmt.Print("\033[1m_ ")
			}
			fmt.Println("\033[m")
		}
		fmt.Print("\033[1m||\033[m")
		for j := range arr[0] {
			// fmt.Print(arr[i][j]," ")		//for print out 0, 1 (real board)
			if arr[i][j] == 0 && i == len(arr)-1 {
				fmt.Print("\033[1m_ \033[m")
			} else {
				fmt.Print(Color[arr[i][j]])
			}
		}
		if i == 1 {
			fmt.Println("\033[1m||          Your score: ", score, "\033[m")
		} else if i == 3 {
			fmt.Println("\033[1m||          Next\033[m")
		} else if i > 4 && tempIndexRow < len(nextBlock.shape) {
			fmt.Print("\033[1m||          \033[m")
			for tempIndexCol < len(nextBlock.shape[0]) {
				fmt.Print(Color[nextBlock.shape[tempIndexRow][tempIndexCol]])
				tempIndexCol++
			}
			fmt.Println("")
			tempIndexRow++
		} else {
			fmt.Println("\033[1m||\033[m")
		}
	}
	fmt.Println("\rPress: W - Rotate, S - Fall down faster, A - Move left, D - Move Right")
	fmt.Println("\rPress: Spacebar - Hard Drop")
	fmt.Println("\rPress: X to exit the program ")
}

//
func readInputKey() string {
	// disable input buffering
	exec.Command("stty", "-F", "/dev/tty", "cbreak", "min", "1").Run()
	// do not display entered characters on the screen
	exec.Command("stty", "-F", "/dev/tty", "-echo").Run()
	var b []byte = make([]byte, 1)
	os.Stdin.Read(b)
	return string(b)
}

//
func threadGetKey(inputQueue chan string) {
	for {
		key := readInputKey()
		if key == "x" || key == "X" {
			exec.Command("stty", "-F", "/dev/tty", "echo").Run()
			os.Exit(0)
		}
		inputQueue <- key
	}
}

//
func clearChan(inputQueue chan string) chan string {
	for len(inputQueue) != 0 {
		<-inputQueue
	}
	return inputQueue
}

//
func doMovement(board [][]int, land [][]int, block Block, nextBlock Block, Color []string, inputQueue chan string, score int, levelUp int) ([][]int, [][]int, Block, chan string, string) {
	var key string
	var preSec int = int(time.Now().UnixNano() / 1000000)
	for len(inputQueue) != 0 {
		key = <-inputQueue
		if key == "w" || key == "W" {
			block.rotate(land)
		} else if key == "a" || key == "A" {
			block.moveLeft(land)
		} else if key == "d" || key == "D" {
			block.moveRight(land)
		} else if key == "s" || key == "S" {
			block.moveDown(land)
		} else if key == " " {
			block.hardDrop(land)
		} else {
			continue
		}
		board = makeCopy(land)
		board = mergeBlock(board, block)
		Display(board, nextBlock, score, Color)
		time.Sleep(time.Microsecond)
		//if time pass (1 sec or less depended on level_up var)
		if (int(time.Now().UnixNano()/1000000)-preSec >= 1000-levelUp) || key == " " {
			break
		}
	}
	return board, land, block, inputQueue, key
}

//Tetris's main game functin
func Tetris(Color []string, Shape [][][]int) {
	//initializing
	land := initBoard()
	board := initBoard()
	score := 0
	//score and time drop
	var levelUp int = int(160 * int(score/500))
	var preSec int = 0
	//
	var key string
	inputQueue := make(chan string, 100)
	//begin get key input thread
	go threadGetKey(inputQueue)
	//begin the game
	block := randomBlock(board, Shape)
	nextBlock := randomBlock(board, Shape)
	board = mergeBlock(board, block)
	Display(board, nextBlock, score, Color)
	//
	for {
		levelUp = int(160 * int(score/500))
		if score >= 3000 {
			levelUp = 900
		}
		//
		board, land, block, inputQueue, key = doMovement(board, land, block, nextBlock, Color, inputQueue, score, levelUp)
		if int(time.Now().UnixNano()/1000000)-preSec >= 1000-levelUp || key == " " {
			block.y++
			if checkCollision(land, block) {
				block.y--
				inputQueue = clearChan(inputQueue)
				land = makeCopy(board)
				land, score = handleFullLine(land, score)
				block = nextBlock
				nextBlock = randomBlock(board, Shape)
			}
			board = makeCopy(land)
			board = mergeBlock(board, block)
			Display(board, nextBlock, score, Color)
			preSec = int(time.Now().UnixNano() / 1000000)
		}
		if checkReachTop(land) {
			exec.Command("stty", "-F", "/dev/tty", "echo").Run()
			fmt.Println("\r\033[1;31mGAME OVER\033[m")
			os.Exit(0)
		} else {
			time.Sleep(time.Millisecond)
		}
	}
}

//Main
func main() {
	rand.Seed(time.Now().UnixNano())
	Tetris(colorCode, blockShape)
}
